projects/dsp-ui/README.md

```diff
@@ -1,26 +1,288 @@
+      const dspApiConfig: KnoraApiConfig = window['tempConfigStorage'] as KnoraApiConfig;
+
+      // init dsp-api configuration
+      AppInitService.dspApiConfig = new KnoraApiConfig(
+        dspApiConfig.apiProtocol,
+        dspApiConfig.apiHost,
+        dspApiConfig.apiPort,
+        dspApiConfig.apiPath,
+        dspApiConfig.jsonWebToken,
+        dspApiConfig.logErrors
+      );
+
+      // set knora-api connection configuration
+      AppInitService.dspApiConnection = new KnoraApiConnection(AppInitService.dspApiConfig);
+
+      resolve();
+    });
+  }
+}
+```
+
+Once we have this service, we have to edit the `app.module.ts` with a function to load the config:
+
+```typescript
+export function initializeApp(appInitService: AppInitService) {
+  return (): Promise<any> => {
+    return appInitService.Init();
+  };
+}
+```
+
+Provide it in the main module and include the desired DSP-UI modules in the imports:
+
+```typescript
+@NgModule({
+  declarations: [
+    AppComponent,
+    ProjectsComponent
+  ],
+  imports: [
+    BrowserModule,
+    AppRoutingModule,
+    DspCoreModule,      // <-- add the dsp-ui modules here
+    DspViewerModule
+  ],
+  providers: [
+      // add the following code block
+    AppInitService,
+    {
+      provide: APP_INITIALIZER,
+      useFactory: initializeApp,
+      deps: [AppInitService],
+      multi: true
+    },
+    {
+      provide: DspApiConfigToken,
+      useFactory: () => AppInitService.dspApiConfig
+    },
+    {
+      provide: DspApiConnectionToken,
+      useFactory: () => AppInitService.dspApiConnection
+    }
+  ],
+  bootstrap: [AppComponent]
+})
+export class AppModule { }
+```
+
+Do not forget to make the imports for `APP_INITIALIZER` from `@angular/core` and the desired DSP-UI modules from `@dasch-swiss/dsp-ui`.
+
+Finally, the `main.ts` file must be modified:
+
+```typescript
+import { enableProdMode } from '@angular/core';
+import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
+import { AppModule } from './app/app.module';
+import { environment } from './environments/environment';
+
+if (environment.production) {
+  enableProdMode();
+}
+
+function bootstrapFailed(result) {
+  console.error('bootstrap-fail! Config is missing or api parameters are not defined', result);
+}
+
+fetch(`config/config.${environment.name}.json`)
+  .then(response => response.json())
+  .then(config => {
+    if (!config) {
+      bootstrapFailed(config);
+      return;
+    }
+
+    // store the response somewhere that the AppInitService can read it.
+    window['tempConfigStorage'] = config;
+
+    platformBrowserDynamic()
+      .bootstrapModule(AppModule)
+      .catch(err => bootstrapFailed(err));
+  })
+  .catch(bootstrapFailed);
+
+```
+
+## Usage
+<!-- TODO: add the modules to app.modules and use them as usual  -->
+<!-- app.modules -->
+Add the desired modules from DSP-UI to the `app.module.ts`. At least `DspCoreModule` has to be imported!
```

Comment:
Actually, the injection tokens are not event part of the Angular module, they are just part of the npm package.

Reply:
I do not understand...

Reply:
the core module as an Angular module does not contain anything:

```ts
@NgModule({
    declarations: [],
    imports: [],
    exports: []
})
export class DspCoreModule {
}
```

The injection tokens are part of the npm package, not the Angular core module.

Reply:
this is an npm export (`index.ts`, `public-api.ts`)

Reply:
meaning that at the moment you can remove `DspCoreModule` from the imports in `app.module.ts` and it will still work as before.

Reply:
That's not true. The core module handles the InjectionToken from the app-init.service. This is already implemented. Otherwise the section `Setup` in the lib README wouldn't make sense...

Reply:
The core.module exports the DspApiConfigToken and DspApiConnectionToken:

```typescript
// config for knora-api-js-lib (@knora/api) config object
export const DspApiConfigToken = new InjectionToken<KnoraApiConfig>('DSP api configuration');

// connection config for knora-api-js-lib (@knora/api) connection
export const DspApiConnectionToken = new InjectionToken<KnoraApiConnection>('DSP api connection instance');
```

Reply:
yes, but this is EcmaScript, not Angular

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export